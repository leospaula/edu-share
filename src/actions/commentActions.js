const BASE_URL = 'https://edusource-api.herokuapp.com/resources'

function addComment(comment){
  const configObj = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json'
    },
    body: JSON.stringify(comment)
  }

  return dispatch => {
    dispatch({type: "ADD_COMMENT", comment})
    fetch(`${BASE_URL}/${comment.resourceId}/comments`, configObj)
  }
}

function fetchComments(resourceId){
  return dispatch => {
    dispatch({ type: "LOAD_COMMENTS", resourceId })
    fetch(BASE_URL + `/${resourceId}/comments`)
      .then(resp => resp.json())
      .then(comments => dispatch({ type: "ADD_COMMENTS", comments }))
  }
}

export { addComment, fetchComments }