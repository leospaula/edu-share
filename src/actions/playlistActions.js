import * as Cookies from "js-cookie"
const BASE_URL = 'https://edusource-api.herokuapp.com/playlists'

const token = () => Cookies.get("eduResourceSession")

function fetchPlaylists(){
  const configObj = {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: token()
    }
  }
  return dispatch => {
    dispatch({type: 'LOAD_PLAYLISTS'})
    fetch(BASE_URL, configObj)
      .then(resp => resp.json())
      .then(playlists => dispatch({
        type: "ADD_PLAYLISTS",
        playlists
      }))
  }
}

function addResourceToPlaylist(resourceID, playlistId){
  const configObj = {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: token()
    },
    body: JSON.stringify({
      update: "add resource",
      resource_id: resourceID,
      id: playlistId
    })
  }
  return dispatch => {
    dispatch({type: "START_ADD", tempPlaylist: {name: "temp", resources: [{id: resourceID}]}})
    fetch(BASE_URL + `/${playlistId}`, configObj)
      .then(resp => resp.json())
      .then(updatedPlaylist => dispatch({
        type: "ADD_RESOURCE_TO_PLAYLIST",
        updatedPlaylist
      }))
  }
}

function removeResourceFromPlaylist(resourceId, playlistId){
  const configObj = {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: token()
    },
    body: JSON.stringify({
      update: "remove resource",
      resource_id: resourceId,
      id: playlistId
    })
  }
  return dispatch => {
    dispatch({
      type: "REMOVE_RESOURCE_FROM_PLAYLIST",
      resourceId,
      playlistId
    })
    fetch(BASE_URL + `/${playlistId}`, configObj)
  }
}

function addPlaylist(playlist){
  const configObj = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: token()
    },
    body: JSON.stringify(playlist)
  }

  return (dispatch) => {
    dispatch({type: "START_ADD", tempPlaylist: playlist})
    fetch(BASE_URL, configObj)
      .then(resp => resp.json())
      .then(respPlaylist => dispatch({
        type: "ADD_PLAYLIST",
        respPlaylist
      }))
  }
}

export { fetchPlaylists, addResourceToPlaylist, addPlaylist, removeResourceFromPlaylist }