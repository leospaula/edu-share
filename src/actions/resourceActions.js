const BASE_URL = 'https://edusource-api.herokuapp.com/resources'

function addResource(resource){
  const configObj = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json'
    },
    body: JSON.stringify(resource)
  }

  return (dispatch) => {
    dispatch({
      type: "ADD_RESOURCE",
      resource
    })
    fetch(BASE_URL, configObj)
  }
}

function fetchResources(subject="all"){
  return (dispatch) => {
    dispatch({type: 'LOAD_RESOURCES'})
    fetch(BASE_URL + `/?q=${subject}`)
      .then(resp => resp.json())
      .then(resources => dispatch({
        type: "ADD_RESOURCES",
        resources
      }))
  }
}

export { addResource, fetchResources }