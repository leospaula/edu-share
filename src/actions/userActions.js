import * as Cookies from "js-cookie"
const BASE_URL = 'https://edusource-api.herokuapp.com'

function buildPostObj(user){
  return ({
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json'
    },
    credentials: 'include',
    body: JSON.stringify(user)
  })
}

function validateUser(dispatch, userData){
  if (userData.valid === "true"){
    Cookies.remove("eduResourceSession")
    Cookies.set("eduResourceSession", userData.token, { expires: 14 })
    dispatch({type: "LOGIN_USER", user: userData.user})
  } else {
    dispatch({type: "INVALID_USER", errors: userData.errorMessages})
  }
  dispatch({type: "COMPLETE_AUTH"})
}

function createUser(user){
  const configObj = buildPostObj(user)

  return dispatch => {
    dispatch({type: "START_AUTH"})
    fetch(BASE_URL.concat('/users'), configObj)
      .then(resp=>resp.json())
      .then(userData => validateUser(dispatch, userData))
  }
}

function loginUser(user){
  const configObj = buildPostObj(user)

  return dispatch => {
    dispatch({type: "START_AUTH"})
    fetch(BASE_URL.concat('/login'), configObj)
      .then(resp=>resp.json())
      .then(userData => validateUser(dispatch, userData))
  }
}

function logoutUser(){
  Cookies.remove("eduResourceSession")
  return dispatch => {
    dispatch({type: "INVALID_USER", errors: "user logged out"})
  }
}

function authorizeUser(){
  let token = Cookies.get("eduResourceSession");
  return dispatch => {
    dispatch({type: "START_AUTH"})
    if (token){
      const configObj = {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Accept: 'application/json',
          Authorization: token
        },
        credentials: 'include'
      }
      fetch(BASE_URL.concat('/authorize'), configObj)
        .then(resp => resp.json())
        .then(authResp => {
          if (authResp.valid === "true"){
            dispatch({type: "LOGIN_USER", user: authResp.user})
          } else {
            dispatch({type: "INVALID_USER", errors: {session: "Please login to continue"}})
          }
          dispatch({type: "COMPLETE_AUTH"})
        })
    } else {
      dispatch({type: "INVALID_USER", errors: {session: "Please login to continue"}})
      dispatch({type: "COMPLETE_AUTH"})
    }
    
  }
}

export { createUser, loginUser, authorizeUser, logoutUser }