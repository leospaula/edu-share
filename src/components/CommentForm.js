import React from 'react'

class CommentForm extends React.Component{
  state = {
    text: ""
  }
  
  render(){
    return(
      <form>
        <textarea name="text" value={this.state.text} placeholder="write comment here..."/ ><br/>
        <input type="submit" value="Post Comment"/>
      </form>
    )
  }
}

export default CommentForm