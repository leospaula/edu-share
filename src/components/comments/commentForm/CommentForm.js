import React from 'react'
import { Link, Redirect } from 'react-router-dom'
import { handleInputChange, handleSubmit } from '../../../utilities'

class CommentForm extends React.Component{
  state = {
    text: "",
    submitted: false
  }

  render(){
    if (this.state.submitted){
      return <Redirect to={`/resources/${this.props.resourceId}`}/>
    } else {
      return(
        <div data-testid="comment-form" className="modal" id="comment-form" tabIndex="-1" role="dialog">
          <div className="modal-dialog modal-dialog-centered" role="document">
            <div className="modal-content">
              <div className="modal-header">
                <h5 className="modal-title">Add Comment</h5>
                <Link data-testid="exit-comment-modal" to={`/resources/${this.props.resourceId}`} type="button" className="close" data-dismiss="modal" aria-label="Close">
                  <span aria-hidden="true">&times;</span>
                </Link>
              </div>
              <div className="modal-body">
                <form onSubmit={e => handleSubmit.call(this, {
                  e,
                  callback: this.props.addComment,
                  currentState: {
                    ...this.state, 
                    userId: this.props.user.current.id, 
                    resourceId: this.props.resourceId,
                    user: this.props.user
                  },
                  clearState: {text: ""}
                })}>
                  <div className="form-group">
                    <textarea className="form-control" name="text" value={this.state.text} placeholder="write comment here..." onChange={e => handleInputChange.call(this, e)} />
                  </div>
                  <input data-testid="comment-submit" className="btn btn-primary tertiary-background" type="submit" value="Post Comment"/>
                </form>
              </div>
            </div>
          </div>
        </div>
      )
    }
  }
}

export default CommentForm