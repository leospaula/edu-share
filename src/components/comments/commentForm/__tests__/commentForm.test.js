import React from 'react'
import ReactDOM from 'react-dom'
import CommentForm from '../CommentForm'
import { BrowserRouter as Router } from 'react-router-dom'
import { addComment } from '../../../../actions/commentActions'
import renderer from 'react-test-renderer'
import {cleanup, fireEvent, render} from '@testing-library/react'

const user = {
  current: {
    id: 1,
    username: 'Charlie'
  },
  valid: true,
  authCompleted: true,
  errors: {}
}

it ('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Router><CommentForm 
    resourceId={1} 
    user={user}
    addComment={addComment}/></Router>, div)
})

it ('matches the snapshot', () => {
  const tree = renderer.create(<Router><CommentForm 
    resourceId={1} 
    user={user}
    addComment={addComment}/></Router>
    ).toJSON()
  expect(tree).toMatchSnapshot()
})

it ('redirects to the resource view once submitted', () => {
  const { getByTestId } = render(
    <Router><CommentForm 
    resourceId={1} 
    user={user}
    addComment={addComment}/></Router>,
  );

  fireEvent.click(getByTestId("comment-submit"));
})

it ('redirects to the resource view if the x is clicked', () => {

})

it ('triggers the addComment action upon submission', () => {
//mock addcomment
//use expect.toBeCalled(mock)
})
