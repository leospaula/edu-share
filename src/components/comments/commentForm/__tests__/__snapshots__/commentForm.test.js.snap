// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div
  className="modal"
  id="comment-form"
  role="dialog"
  tabIndex="-1"
>
  <div
    className="modal-dialog modal-dialog-centered"
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h5
          className="modal-title"
        >
          Add Comment
        </h5>
        <a
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          data-testid="exit-comment-modal"
          href="/resources/1"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </a>
      </div>
      <div
        className="modal-body"
      >
        <form
          onSubmit={[Function]}
        >
          <div
            className="form-group"
          >
            <textarea
              className="form-control"
              name="text"
              onChange={[Function]}
              placeholder="write comment here..."
              value=""
            />
          </div>
          <input
            className="btn btn-primary tertiary-background"
            data-testid="comment-submit"
            type="submit"
            value="Post Comment"
          />
        </form>
      </div>
    </div>
  </div>
</div>
`;
